File: web/src/repo/wal-monitor.ts
import { api } from './client'

export interface WALMonitor {
    id: number
    source_id: number
    source?: {
        name: string
        [key: string]: any
    }
    wal_lsn: string | null
    wal_position: number | null
    last_wal_received: string | null
    last_transaction_time: string | null
    replication_slot_name: string | null
    replication_lag_bytes: number | null
    total_wal_size: string | null
    status: 'ACTIVE' | 'IDLE' | 'ERROR'
    error_message: string | null
    updated_at: string
}

export interface WALMonitorListResponse {
    monitors: WALMonitor[]
    total: number
}

export const walMonitorRepo = {
    getAll: async () => {
        const { data } = await api.get<WALMonitorListResponse>('/wal-monitor')
        return data
    },
}

File: backend/app/domain/schemas/wal_monitor.py
"""
WAL Monitor schemas for API requests and responses.

Pydantic models for validation and serialization.
"""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class WALMonitorBase(BaseModel):
    """Base schema for WAL monitor with common fields."""

    wal_lsn: Optional[str] = Field(None, description="Log Sequence Number")
    wal_position: Optional[int] = Field(
        None, description="WAL position as numeric value"
    )
    last_wal_received: Optional[datetime] = Field(
        None, description="Last WAL receive time"
    )
    last_transaction_time: Optional[datetime] = Field(
        None, description="Last transaction timestamp"
    )
    replication_slot_name: Optional[str] = Field(
        None, description="Replication slot name"
    )
    replication_lag_bytes: Optional[int] = Field(
        None, description="Replication lag in bytes"
    )
    status: str = Field(default="ACTIVE", description="Status: ACTIVE, IDLE, ERROR")
    error_message: Optional[str] = Field(None, description="Error details if any")


class WALMonitorCreate(WALMonitorBase):
    """Schema for creating WAL monitor record."""

    source_id: int = Field(..., description="Source identifier")


class WALMonitorUpdate(WALMonitorBase):
    """Schema for updating WAL monitor record (all fields optional)."""

    pass


class WALMonitorStatusUpdate(BaseModel):
    """Schema for updating only status."""

    status: str = Field(..., description="Status: ACTIVE, IDLE, ERROR")
    error_message: Optional[str] = Field(
        None, description="Error message if status is ERROR"
    )


class WALMonitorResponse(WALMonitorBase):
    """Schema for WAL monitor response."""

    id: int = Field(..., description="Monitor record ID")
    source_id: int = Field(..., description="Source identifier")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    class Config:
        orm_mode = True


class WALMonitorListResponse(BaseModel):
    """Schema for list of WAL monitors."""

    monitors: List[WALMonitorResponse] = Field(
        ..., description="List of WAL monitor records"
    )
    total: int = Field(..., description="Total number of records")
